crates/bevy_ecs/src/query/par_iter.rs

```diff
@@ -0,0 +1,135 @@
+    /// that is being initialized and run from the ECS scheduler, this should never panic.
+    ///
+    /// [`ComputeTaskPool`]: bevy_tasks::ComputeTaskPool
+    #[inline]
+    pub fn for_each_mut<FN: Fn(QueryItem<'w, Q>) + Send + Sync + Clone>(self, func: FN) {
+        let batch_size = match self.batch_size.or_else(|| self.get_default_batch_size()) {
+            Some(batch_size) => batch_size.max(1),
+            None => return,
+        };
+        // SAFETY: query has unique world access
+        unsafe {
+            self.state
+                .par_for_each_unchecked_manual::<QueryFetch<Q>, FN>(
+                    self.world,
+                    batch_size,
+                    func,
+                    self.world.last_change_tick(),
+                    self.world.read_change_tick(),
+                );
+        }
+    }
+
+    /// Runs `func` on each query result in parallel.
+    ///
+    /// # Panics
+    /// The [`ComputeTaskPool`] resource must be added to the `World` before using this method. If using this from a query
+    /// that is being initialized and run from the ECS scheduler, this should never panic.
+    ///
+    /// # Safety
+    ///
+    /// This does not check for mutable query correctness. To be safe, make sure mutable queries
+    /// have unique access to the components they query.
+    ///
+    /// [`ComputeTaskPool`]: bevy_tasks::ComputeTaskPool
+    #[inline]
+    pub unsafe fn for_each_unchecked<FN: Fn(QueryItem<'w, Q>) + Send + Sync + Clone>(
+        self,
+        func: FN,
+    ) {
+        let batch_size = match self.batch_size.or_else(|| self.get_default_batch_size()) {
+            Some(batch_size) => batch_size.max(1),
+            None => return,
+        };
+        self.state
+            .par_for_each_unchecked_manual::<QueryFetch<Q>, FN>(
+                self.world,
+                batch_size,
+                func,
+                self.world.last_change_tick(),
+                self.world.read_change_tick(),
+            );
+    }
+
+    fn get_default_batch_size(&self) -> Option<usize> {
+        let thread_count = self
+            .state
+            .task_pool
+            .as_ref()
+            .map(|pool| pool.thread_num())
+            .unwrap_or(0);
+        assert!(
+            thread_count > 0,
+            "Attempted to run parallel iteration over a query with an empty TaskPool"
+        );
+        let max_size = if QF::IS_DENSE && <QueryFetch<'static, F>>::IS_DENSE {
+            let tables = &self.world.storages().tables;
+            self.state
+                .matched_table_ids
+                .iter()
+                .map(|id| tables[*id].len())
+                .max()
+        } else {
+            let archetypes = &self.world.archetypes();
+            self.state
+                .matched_archetype_ids
+                .iter()
+                .map(|id| archetypes[*id].len())
+                .max()
+        };
+        max_size.map(|max| max / thread_count)
```

Comment:
I think max / thread_count is too coarse.  Consider the case where we have one archetype, if for any reason  we don't get all the threads, the task will take twice as long to complete. This is because one of the threads will have to do 2 batches instead of just one. We probably want some type of multiplier on the thread count so we get like 4 or 5 batches per a thread. Probably needs some testing to know what that number is.

Reply:
Set one up at a default of 4 per thread for now. Ultimately it comes down to how much of a cost the overhead of scheduling a new task onto a thread is. If that overhead is small, we can afford to be much more granular with how we schedule each one.

As I proposed on Discord, it might be a good idea to schedule one task per thread, and just push the archetypes/batches through a SPMC channel. It'd be likely be lighter weight scheduling new tasks.